modules/core/src/main/java/org/apache/ignite/internal/manager/RECOVERY.md

```diff
@@ -0,0 +1,89 @@
+## Local state recovery
+
+### Definitions
+
+**Local state** is a state of components, started tables, schemas, persistent and in-memory storages, partitions, RAFT groups, etc. on a given node. It also includes a projection of distributed cluster configuration.
+
+**Local state recovery** is changing local state in a way that every component on the node has its state configured and updated accordingly to cluster-wide configuration, topology, stored data, etc.
+
+**Vault** is a local storage of node configuration and a projection of cluster configuration. Vault has no log of historical changes. See [Vault](../../../../../../../../../vault/README.md).
+
+**Metastorage** is a distributed storage of cluster-wide meta information, including cluster [configuration](../../../../../../../../../configuration/README.md). Metastorage is located on a group of nodes, forming a RAFT group. It stores a history of changes.
+
+**CMG, or Cluster Management group**, is responsible for coordinating node join, node left, logical topology, etc.
+
+**Distributed phase of recovery** is a phase when the node retrieves updates from Metastorage and applies them to make its state up-to-date with the cluster. It is required when the node is trying to join the cluster, and is not, when it starts in standalone mode, for example, in maintenance mode.
+
+**Physical topology** - a set of nodes that are currently online and discovered via SWIM protocol, see [network-api](../../../../../../../../../network-api/README.md) and [network](../../../../../../../../../network/README.md) for details.
+
+**Logical topology** - a set of nodes that are present in physical topology and finished recovery process. They are considered to be able to serve user load.
+
+### Prerequisites
+
+1. Vault is either empty or contains data that is available for reading (no corruption).
+   If any action is required to recover data in Vault, it should be executed before.
+2. Data in Vault is consistent to the moment when the node was stopped (last applied revision appropriates to the rest of the data).
+3. Vault contains local node configuration, and a projection of cluster configuration. The latter can be written only after the cluster is initialized. No conflicts are possible between the node local configuration and the cluster configuration that is stored in Metastorage because the properties’ sets don’t intersect with each other.
+4. The prerequisite for the recovery from Metastorage is receiving a successful _nodeJoinResponse_, see [Node join protocol](TODO add readme).
```

Comment:
Do we have an issue for this TODO? As I know, we can't merge any TODOs without issue number.